### ISSUE TITLE
Sherpa tests with Python 3.6

### ISSUE THREAD

**Issue by cdeil** [sherpa/sherpa#360]
Hi, I updated from Python 3.5 to 3.6. Everything works OK (all Gammapy tests pass). There's just a few things that I noticed during build and when running `sherpa_test` that I thought you might want to know about. Feel free to close this issue any time if those aren't relevant.

First I got this: https://gist.github.com/cdeil/94e3f758e5501db7fc014a402accbdcb
I guess `mock` could be mentioned here?
https://github.com/sherpa/sherpa#testing-sherpa

After installing mock I got this: https://gist.github.com/cdeil/0306b8e16954599bdab18574997733d8
I guess there's a few missing `requires_data` decorators in your tests?
Shouldn't this be caught by your CI?

During install this is printed in red. Is is normal?
```
zip_safe flag not set; analyzing archive contents...
sherpa.__pycache__.__init__.cpython-36: module references __file__
sherpa.__pycache__._version.cpython-36: module references __file__
sherpa.astro.datastack.tests.__pycache__.test_datastack.cpython-36: module references __file__
sherpa.astro.models.__pycache__._modelfcts.cpython-36: module references __file__
sherpa.astro.ui.__pycache__.serialize.cpython-36: module MAY be using inspect.getsource
sherpa.astro.ui.tests.__pycache__.test_astro_ui.cpython-36: module references __file__
sherpa.astro.utils.__pycache__._pileup.cpython-36: module references __file__
sherpa.astro.utils.__pycache__._region.cpython-36: module references __file__
sherpa.astro.utils.__pycache__._utils.cpython-36: module references __file__
sherpa.astro.utils.__pycache__._wcs.cpython-36: module references __file__
sherpa.astro.utils.__pycache__.smoke.cpython-36: module references __file__
sherpa.estmethods.__pycache__._est_funcs.cpython-36: module references __file__
sherpa.models.__pycache__._modelfcts.cpython-36: module references __file__
sherpa.optmethods.__pycache__._minim.cpython-36: module references __file__
sherpa.optmethods.__pycache__._minpack.cpython-36: module references __file__
sherpa.optmethods.__pycache__._saoopt.cpython-36: module references __file__
sherpa.optmethods.__pycache__._tstoptfct.cpython-36: module references __file__
sherpa.stats.__pycache__._statfcts.cpython-36: module references __file__
sherpa.tests.__pycache__.test_stack.cpython-36: module references __file__
sherpa.utils.__pycache__.__init__.cpython-36: module references __file__
sherpa.utils.__pycache__._psf.cpython-36: module references __file__
sherpa.utils.__pycache__._utils.cpython-36: module references __file__
sherpa.utils.__pycache__.integration.cpython-36: module references __file__
```
This is my full build log: https://gist.github.com/cdeil/c646e6f7b15313590c418dae70d16aa9

For me, the version has an egg in it:
```python
>>> import sherpa
>>> sherpa.__version__
'4.9.0+28.gaa98d19.dirty-py3.6-macosx-10.12-x86_64.egg'
```
Is it normal / expected?

**Comment by DougBurke** [comment]
We currently don't have the resources to focus on Python 3.6, so any input from the community would be helpful. The `module references __file__` message looks like it may be down to `easy_install` - e.g. http://stackoverflow.com/questions/2298403/warnings-emitted-during-easy-install - but I admit to glazing over at this part of the python ecosystem.

We are trying to avoid too many test builds on Travis, since we're not paying for it, and so don't have a case for every possible combination of decorator. I don't have time to look at your install closely: are you running in a case where you don't have the external data installed but do have astropy installed (so do have a FITS reader), or do you have neither? Perhaps we could add a test of this option.

**Comment by DougBurke** [comment]
@cdeil would it be possible for you to test out the current `master` branch to see if things are any better for you with Python 3.6. We now have a minimal test on Travis to catch those tests which rely on "extra" features (e.g. the test data) which had tripped you up, and the README.md has seen a "very minimal" addition to the testing section.

**Comment by olaurino** [comment]
I couldn't have an opportunity to reproduce this issue, but to the best of my understanding the issue is that we should not import `mock` like this (`import mock`) on Python 3, where the library is already available as part of the standard library (see #317, but for a different offending file in that case). On the other hand, `mock` is indeed also available as a standalone package in Python 3, and `sherpa_test` tries to install it when it doesn't find it. That would require permissions to write in the Python installation directory, which is not the case in @cdeil's setup. I wonder if the installation could be done by means other than `pip` in `sherpa.__init__.py`, or maybe `pip` should be called with options to perform a user-only installation.

I've been trying to cover all possibilities, but corner cases still pop up, and admittedly I tend to assume people use virtual environments or Anaconda rather than relying on the system's Python, which is however a valid setup too.

Any changes to this logic would require setting up additional CI tests that do not rely on `conda` or other kinds of isolated, user-level environments.

**Comment by olaurino** [comment]
@cdeil I don't think this issue is relevant anymore. We fixed the `mock` imports (#420), we have been supporting Python 3.6 for a while, and the missing decorators were added (#363). Your use case is still kind of outside of our regular parameters as we don't support hacking "in place", you need to either `develop` or `install`, but other than that the reported issues should all have been addressed.

I am closing the issue, but please feel free to let me know if I should reopen it.
